### ISSUE TITLE
Rewriting of projections

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#1981]
Andy Pitts reports that he would like to rewrite projections:

``` agda
{-# OPTIONS --rewriting #-}

infix 4 _↦_
postulate
  _↦_ : {A : Set} → A → A → Set

{-# BUILTIN REWRITE _↦_ #-}

postulate
  𝔸 : Set
  sw : 𝔸 → 𝔸 → 𝔸 → 𝔸

record Act (A : Set) : Set where
  field
    swap : 𝔸 → 𝔸 → A → A

    swap-id :
      (a : 𝔸)
      (x : A)
      → -----------
      swap a a x ↦ x
open Act  public

type-of-swap-id : ∀ {A} (r : Act A) (a : 𝔸) x → swap r a a x ↦ x
type-of-swap-id = swap-id

{-# REWRITE swap-id #-}

test-swap-id : ∀ {A} {P : A → Set} (r : Act A) (a : 𝔸) x → P (swap r a a x) → P x
test-swap-id r a x h = h
```

Currently, rewriting of projections is not supported, since terms are in spine form.

**Comment by andreasabel** [comment]
Due to our spine form, the following is allowed

``` agda
postulate
  act : Act 𝔸
  act-swap : swap act ↦ sw

{-# REWRITE act-swap #-}
```

and counts as a rewrite rule for `act` (not for `swap`)! 
Currently, rewrite rules are organized by head symbol wrt. spine form.  If we allow rewriting with projections, we have to consider different heads of a term, e.g., in `swap act`, which is internally `act. swap`, both rewrite rules for `act` and `swap` have to be considered.

**Comment by amp12** [comment]
OK. So with rewriting as it is currently, I could get round the problem, case-by-case, by using data types instead of record types. For example

``` agda
    {-# OPTIONS --rewriting #-}

    infix 4 _↦_
    postulate
      _↦_ : {A : Set} → A → A → Set

    {-# BUILTIN REWRITE _↦_ #-}

    data ActData (A : Set) : Set where
      act :
        (swap : 𝔸 → 𝔸 → A → A) 
        (swap-id : (a : 𝔸)(x : A) → swap a a x ↦ x)
        → -----------------------------------------
        ActData A

    prs : {A : Set} → ActData A →  𝔸 → 𝔸 → A → A
    prs (act s i) = s

    pri :  {A : Set}(r : ActData A)(a : 𝔸)(x : A) → prs r a a x ↦ x
    pri (act s i) = i

    {-# REWRITE pri #-}

    test : ∀ {A} {P : A → Set} (r : ActData A) (a : 𝔸) x → P (prs r a a x) → P x
    test r a x h = h
```

**Comment by andreasabel** [comment]
Indeed, if it not too much of a hassle, a workaround seems best at the moment.
You can keep the rewrite rules in the record if you want.

``` agda
data Act (A : Set) : Set where
  actData : (swap :  𝔸 → 𝔸 → A → A) → Act A

swap : ∀{A} → Act A →  𝔸 → 𝔸 → A → A
swap (actData s) = s

record IsAct (A : Set) (act : Act A) : Set where
  field
    swap-id :
      (a : 𝔸)
      (x : A)
      → -----------
      swap act a a x ↦ x

open IsAct  public

{-# REWRITE swap-id #-}

test-swap-id : ∀ {A} {P : A → Set} (r : Act A) (a : 𝔸) x → P (swap r a a x) → P x
test-swap-id r a x h = h
```

**Comment by andreasabel** [comment]
Why is this a `wontfix`?

**Comment by UlfNorell** [comment]
I interpreted the discussion as *"it'd be really awkward to allow rewrite rules with variable heads and there's a reasonable workaround"*. If you think this is something we'd like to allow I have no objections.

**Comment by jespercockx** [comment]
I'm considering to give this a try, but change a bit how rewrite rules on projections would work. In particular, I propose that we should be able to declare rewrite rules *inside* a record type. Here's an example of what I have in mind:

```agda
open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Agda.Builtin.Nat

variable k l m n : Nat

cong : {A B : Set} (f : A → B) {x y : A} → x ≡ y → f x ≡ f y
cong f refl = refl

record ParallelPlus : Set where
  field
    plus : Nat → Nat → Nat
    plus-0l : plus 0 l ≡ l 
    plus-0r : plus k 0 ≡ k
    plus-sucl : plus (suc k) l ≡ suc (plus k l) 
    plus-sucr : plus k (suc l) ≡ suc (plus k l) 

  {-# REWRITE plus-0l plus-0r plus-sucl plus-sucr #-}

module M (pp : ParallelPlus) where

  open ParallelPlus pp

  plus-comm : plus k l ≡ plus l k
  plus-comm {zero} {l} = refl
  plus-comm {suc k} {l} = cong suc (plus-comm {k} {l})
```
When actually constructing an element of the record type, the equalities should then be checked to hold definitionally, in order to avoid problems with subject reduction.

**Comment by amp12** [comment]
Hi Jesper. That looks very nice — I would use it. At the moment I have to go through an annoying yoga of declaring single-constructor data types instead of records, defining functions on the data that would have been the fields I wanted, and only then using the existing REWRITE facility; the code looks ugly.

Andy

**Comment by jespercockx** [comment]
I'm trying again to make this work and I've been able to make some progress on the internal errors that were occurring in my previous attempt. However, now I'm stuck on a rather more substantial issue: 

- Rewrite rules need to respect eta-equality and irrelevance
- Hence, the matching algorithm for applying a rewrite rule needs to be type-directed
- Hence, rewriting needs to know the type of the head symbol when applying a rewrite rule
- When the head symbol is a projection, its type depends on the parameters of the record type
- To determine the parameters, we need to know the type of the principal argument (i.e. the record value)
- In particular, in case the principal argument is (an application of) a variable, we need to know its type 
- However, currently Agda often reduces a term while in a wrong context

So to make rewrite rules on projections work, we should change Agda so it only calls reduction on a term that is actually in right context.

I've been working on fixing this and added quite a few calls to `addContext` in places, but now I've encountered another one that's trickier to deal with: the occurs check. The problem is that this check traverses terms (going under lambdas) without knowing the type, so knowing the type of variables is not possible. The best solution would be to make the occurs check (and possibly other traversals) type-directed, though I'm afraid this could take quite some work.

**Comment by Alizter** [comment]
@jespercockx Any progress with this?

**Comment by mikeshulman** [comment]
Is this likely to ever be implemented?  I think it would be really nice to have rewriting on projections.  In particular:

1. Because I want to rewrite on their "projections", my Σ-types currently can't be records.  For a while I was making them just postulates, but then their constructors and projections must carry around their parameters all the time, which slows things down a lot.  So now I'm making them datatypes, so that at least their constructors don't need parameters; but this has the drawback that they can't have eta-rules.  (With postulates I could implement eta-contraction with a rewrite, but not with datatypes after #5996.)  But if we could rewrite on projections, then I could use records, which would presumably be both faster and have eta-rules.
2. I also have a situation where (I believe) a rewrite rule leads to nontermination _because_ of the parameters in a projection.  Specifically, the RHS of a rewrite rule is a projection, one of whose parameters contains an instance of the LHS of that rule.  So when it is normalized, the rewrite rule fires inside the parameter, producing another instance of the RHS which contains another parameter, and so on.  But if I could use records, I hope that maybe this wouldn't happen because the projection wouldn't store its parameters.

**Comment by jespercockx** [comment]
I'd still very much like to implement this, but I currently have other priorities. My WIP can be found on https://github.com/jespercockx/agda/tree/rewriting-projections, if anyone wants pointers on where I got stuck I'd be glad to give them.

**Comment by jespercockx** [comment]
I've rebased my branch at https://github.com/jespercockx/agda/tree/rewriting-projections onto a recent version of Agda and tried to run the tests. Basic tests with rewriting projections work, but there are a lot of errors in the test suite with variables being out of scope, and I do not see an easy way to fix them.
